### PR TITLE
feat: Add GetStructField expression

### LINKED ISSUE #730: Support GetStructField
### What is the problem the feature request solves?

To start supporting StructTypes, GetStructField has to be implemented. Nearly all operations on a struct involve just getting one of the values inside via GetStructField, so implementing this one operator can solve a lot of nested use cases.

### Describe the potential solution

Implement GetStructField expr

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Kimahriman**
## Which issue does this PR close?

Closes #730 

## Rationale for this change

To support struct types in expressions, you need to be able to pull out values from structs, which Spark does through the GetStructField expression.

## What changes are included in this PR?

Adds a new PhysicalExpr `GetStructField` that gets a field from within a struct.

Additionally to support this and testing it:
- Creates a new trait `DataTypeSupport` that lets operators opt-in to supporting certain data types. This will help incrementally add complex data type support.
- Updates `CometRowToColumnar` to support columnar sources. This is so I can use Spark's vectorized parquet read to read complex data, and then immediately forward to Comet via `COMET_ROW_TO_COLUMNAR_SUPPORTED_OPERATOR_LIST`. I do this in the UT I made.

## How are these changes tested?

New UT showing comet operators take affect, existing Spark tests for functionality

### CODE REVIEW

File: spark/src/main/scala/org/apache/spark/sql/comet/CometRowToColumnarExec.scala [hunk 1]
```diff
@@ -60,8 +62,17 @@ case class CometRowToColumnarExec(child: SparkPlan)
...
+      child
+        .executeColumnar()
+        .mapPartitionsInternal { iter =>
+          iter.flatMap(_.rowIterator().asScala)
+        }
+    } else {
+      child.execute()
+    }
+
+    rdd
```
Comment (Kimahriman) [inline spark/src/main/scala/org/apache/spark/sql/comet/CometRowToColumnarExec.scala:75]: There might be a more efficient way than using a row iterator to write to the row based arrow writer, but as this is mostly for testing/fallback purposes, I didn't try to figure out a faster Spark vector to arrow vector approach. Hopefully someone is able to add complex type support to the Comet parquet reader. If not it could be worth thinking about using the Spark reader as a real fallback use case instead of just for testing purposes.

**Review by Kimahriman — COMMENTED** [review]
Did my best at implementing a new expression. Had to adjust some typing checking around, since nearly all of the type checking is off a single list or two of supported types, but I'm only trying to add some struct support. Got all the tests to pass at least.

File: spark/src/main/scala/org/apache/spark/sql/comet/CometRowToColumnarExec.scala (hunk 1 above)
Comment (Kimahriman) [reply spark/src/main/scala/org/apache/spark/sql/comet/CometRowToColumnarExec.scala:75]: Also not sure if this means the class should be renamed from `RowToColumnar` since this is technically just doing `ColumnarToColumnar`

File: spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala [hunk 2]
```diff
@@ -59,12 +59,13 @@ object QueryPlanSerde extends Logging with ShimQueryPlanSerde with CometExprShim
     logWarning(s"Comet native execution is disabled due to: $reason")
   }
 
-  def supportedDataType(dt: DataType): Boolean = dt match {
+  def supportedDataType(dt: DataType, allowComplex: Boolean = true): Boolean = dt match {
```
[the remark this answers is not in the document]
Comment (Kimahriman) [reply spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala]: Changed the default to false

File: native/core/src/execution/datafusion/expressions/structs.rs [hunk 3]
```diff
@@ -125,3 +125,103 @@ impl PartialEq<dyn Any> for CreateNamedStruct {
             .unwrap_or(false)
     }
 }
+
+#[derive(Debug, Hash)]
+pub struct GetStructField {
```
Comment (andygrove) [inline native/core/src/execution/datafusion/expressions/structs.rs:130]: I think this can move to the `spark-expr` crate?

File: spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala [hunk 4]
```diff
@@ -1128,7 +1133,7 @@ object CometSparkSessionExtensions extends Logging {
     // Only consider converting leaf nodes to columnar currently, so that all the following
     // operators can have a chance to be converted to columnar.
     // TODO: consider converting other intermediate operators to columnar.
-    op.isInstanceOf[LeafExecNode] && !op.supportsColumnar && isSchemaSupported(op.schema) &&
```
Comment (andygrove) [inline spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:1131]: Could you update the comments here to cover this change in functionality?

File: spark/src/main/scala/org/apache/spark/sql/comet/CometRowToColumnarExec.scala [hunk 5]
```diff
@@ -60,8 +62,17 @@ case class CometRowToColumnarExec(child: SparkPlan)
...
-      .execute()
+    val rdd: RDD[InternalRow] = if (child.supportsColumnar) {
+      child
+        .executeColumnar()
+        .mapPartitionsInternal { iter =>
+          iter.flatMap(_.rowIterator().asScala)
+        }
+    } else {
+      child.execute()
+    }
```
Comment (andygrove) [inline spark/src/main/scala/org/apache/spark/sql/comet/CometRowToColumnarExec.scala:73]: My understanding of this is that it allows us to read from parquet files containing structs in tests by falling back to Spark, and that this new code wouldn't be executed outside of tests? Could you add some comments in here to that effect if this is the case?

**Review by andygrove — COMMENTED** [review]
Thanks for the contribution @Kimahriman. I think this looks good.

File: native/core/src/execution/datafusion/expressions/structs.rs (hunk 3 above)
Comment (Kimahriman) [reply native/core/src/execution/datafusion/expressions/structs.rs:130]: Yeah I thought about that, but seemed to make sense to co-locate with CreateNamedStruct. Is it ok to move both in this PR, or do you want me to just move this new one?

File: spark/src/main/scala/org/apache/spark/sql/comet/CometRowToColumnarExec.scala (hunk 5 above)
Comment (Kimahriman) [reply spark/src/main/scala/org/apache/spark/sql/comet/CometRowToColumnarExec.scala:73]: I mean I would say this is no more-or-less just for testing than the existing use cases for CometRowToColumnarExec. I was using this for real queries while testing this out on real data. It could be a legitimate use depending on when complex type support is added to the parquet readers.

File: native/core/src/execution/datafusion/expressions/structs.rs (hunk 3 above)
Comment (andygrove) [reply native/core/src/execution/datafusion/expressions/structs.rs:130]: I agree that CreateNamedStruct should also be in spark-expr, so I think it is fine to move it as part of this PR

File: spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala (hunk 4 above)
Comment (Kimahriman) [reply spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:1131]: Added a little bit more, but still is just a little weird since it's technically still called `RowToColumnar`

File: native/core/src/execution/datafusion/expressions/structs.rs (hunk 3 above)
Comment (Kimahriman) [reply native/core/src/execution/datafusion/expressions/structs.rs]: Migrated both

File: native/spark-expr/src/structs.rs [hunk 6]
```diff
@@ -125,3 +125,103 @@ impl PartialEq<dyn Any> for CreateNamedStruct {
...
+            DataType::Struct(fields) => Ok(fields[self.ordinal].clone()),
+            data_type => Err(DataFusionError::Plan(format!(
+                "Expect struct field, got {:?}",
+                data_type
+            ))),
+        }
+    }
+}
+
+impl PhysicalExpr for GetStructField {
```
[the remark this answers is not in the document]
Comment (Kimahriman) [reply native/spark-expr/src/structs.rs:151]: Don't know enough about DataFusion to really know what the difference is. Just on the Spark side, UDFs are usually slightly less performant, so if you don't have to do a UDF you're usually better off. DataFusion does have a [get_field](https://github.com/apache/datafusion/blob/main/datafusion/functions/src/core/getfield.rs) ScalarUDF already it looks like, but that's by name and not index, and there seems like a lot more ceremony about checking all the input types, vs the PhysicalExpr is more tailored to what we already know about the input data.

[the remark this answers is not in the document]
Comment (andygrove) [reply native/spark-expr/src/structs.rs:151]: I agree that we should start implementing functions as ScalarUDFImpl instead of PhysicalExpr. I think it would be fine to convert this one as a follow on PR.

[the remark this answers is not in the document]
Comment (Kimahriman) [reply native/spark-expr/src/structs.rs:151]: What's the tl;dr on the benefits of that? Easier to use outside of the Comet/Spark use case?

[the remark this answers is not in the document]
Comment (andygrove) [reply native/spark-expr/src/structs.rs:151]: Yes, I think it makes it easier for DataFusion users to switch between different function implementations and can be used from the logical plan as well as from the physical plan

**Review by andygrove — APPROVED** [review]
LGTM. Thanks @Kimahriman

File: spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala [hunk 7]
```diff
@@ -59,12 +59,13 @@ object QueryPlanSerde extends Logging with ShimQueryPlanSerde with CometExprShim
     logWarning(s"Comet native execution is disabled due to: $reason")
   }
 
-  def supportedDataType(dt: DataType): Boolean = dt match {
+  def supportedDataType(dt: DataType, allowComplex: Boolean = false): Boolean = dt match {
```
Comment (comphead) [inline spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:62]: just thinking do we really need `allowComplex`, which more looks like a `supportComplexTypes` or we can use the Comet param like its done for operator/expressions.

Moreover the isSchemaSupported doesn't have this flag which causes some inconsistency imho

File: spark/src/main/scala/org/apache/spark/sql/comet/CometBatchScanExec.scala [hunk 8]
```diff
@@ -152,3 +153,17 @@ case class CometBatchScanExec(wrapped: BatchScanExec, runtimeFilters: Seq[Expres
 
   override def supportsColumnar: Boolean = true
 }
+
+object CometBatchScanExec {
+
+  def isSchemaSupported(schema: StructType): Boolean =
```
Comment (kazuyukitanimura) [inline spark/src/main/scala/org/apache/spark/sql/comet/CometBatchScanExec.scala:159]: What was the reason of copying `isSchemaSupported` and `isTypeSupported` here?
It looks they are still used only in `CometSparkSessionExtensions.scala`, so that we can keep using the original one?
Also there is another copy in `CometScanExec.scala`

File: spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala [hunk 9]
```diff
@@ -1756,4 +1756,21 @@ class CometExpressionSuite extends CometTestBase with AdaptiveSparkPlanHelper {
...
   }
+
+  test("get_struct_field") {
+    withSQLConf(
+      CometConf.COMET_ROW_TO_COLUMNAR_ENABLED.key -> "true",
+      CometConf.COMET_ROW_TO_COLUMNAR_SUPPORTED_OPERATOR_LIST.key -> "FileSourceScan") {
+      withTempPath { dir =>
+        var df = spark
+          .range(5)
+          .select(struct(col("id")).alias("nested"))
```
Comment (kazuyukitanimura) [inline spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala:1767]: Do you mind adding a null row (the struct is null) to this test case?

Comment (Kimahriman) [reply spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala:1767]: Hah I did have that originally but it looked like most things just relied on the Spark tests for extensive testing, I can add back in some nulls

File: spark/src/main/scala/org/apache/spark/sql/comet/CometBatchScanExec.scala (hunk 8 above)
Comment (Kimahriman) [reply spark/src/main/scala/org/apache/spark/sql/comet/CometBatchScanExec.scala:159]: Currently it's very hard coded that everything in Comet supports the exact same data types. This was an attempt to start breaking out supported data types for different functionality within Comet. Just because I want to support struct types in expressions, doesn't mean I have to support it in the Scan operations. And when the Scans are updated to support each complex type, they can then opt-in to supporting them. And CometBatchScanExec and CometScanExec don't need to be updated at the same time to support the same types either.

We could get away with just delegating back to the CometSparkSessionExtensions check with the new allowComplex flag, but it seems to make a lot more sense to let each operator opt-in to exactly what it supports.

File: spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala (hunk 7 above)
Comment (Kimahriman) [reply spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:62]: Yes there's a lot of oddities in how supported data types works right now. It should really just be operator/expression dependent that gets bubbled up to whether the whole plan is supported or not. 

I'm not sure what you mean by 
>or we can use the Comet param like its done for operator/expressions.

File: spark/src/main/scala/org/apache/spark/sql/comet/CometBatchScanExec.scala (hunk 8 above)
Comment (kazuyukitanimura) [reply spark/src/main/scala/org/apache/spark/sql/comet/CometBatchScanExec.scala:159]: Got it. Let's delete the original one then? With this PR it is no longer used I think.

Also this object should be package private unless we add a test for it. `isTypeSupported` should be `private` as well

Comment (Kimahriman) [reply spark/src/main/scala/org/apache/spark/sql/comet/CometBatchScanExec.scala:159]: It's still used just by `shouldApplyRowToColumnar`, but to go with this theme I could move that onto the CometRowToColumnarExec object

Comment (Kimahriman) [reply spark/src/main/scala/org/apache/spark/sql/comet/CometBatchScanExec.scala:159]: I don't think I can make `object CometBatchScanExec` private? Since it's being accessed from `org.apache.comet` and it's in `org.apache.spark.sql.comet`. I think all I can do is make `isTypeSupported` private?

File: spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala (hunk 7 above)
Comment (comphead) [reply spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:62]: Like 
```
  private[comet] def isCometAllOperatorEnabled(conf: SQLConf): Boolean = {
    COMET_EXEC_ALL_OPERATOR_ENABLED.get(conf)
  }
```

File: spark/src/main/scala/org/apache/spark/sql/comet/CometBatchScanExec.scala (hunk 8 above)
Comment (kazuyukitanimura) [reply spark/src/main/scala/org/apache/spark/sql/comet/CometBatchScanExec.scala:159]: I see. Is it possible to create a base trait? So that we can share the logic for be basic types between different objects.
Then perhaps new `CometRowToColumnarExec` object can inherit it and call super + extra type match.

I cannot think cases like a type supported only in `CometBatchScanExec` but not in `CometScanExec`...

File: spark/src/main/scala/org/apache/spark/sql/comet/CometRowToColumnarExec.scala (hunk 1 above)
Comment (andygrove) [reply spark/src/main/scala/org/apache/spark/sql/comet/CometRowToColumnarExec.scala:75]: This class is now capable of converting Spark rows or Spark columns to Comet/Arrow columnar format. Perhaps we should consider renaming it. I'm not sure what a good name would be. Some ideas: 

- CometSparkToColumnarExec
- CometConvertFromSparkExec

Comment (Kimahriman) [reply spark/src/main/scala/org/apache/spark/sql/comet/CometRowToColumnarExec.scala:75]: Yeah I agree a new name would probably make sense. Was worried about bloating this PR too much though 😅

File: spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala [hunk 10]
```diff
@@ -1115,6 +1119,7 @@ object CometSparkSessionExtensions extends Logging {
         BinaryType | StringType | _: DecimalType | DateType | TimestampType =>
       true
     case t: DataType if t.typeName == "timestamp_ntz" => true
+    case s: StructType => isSchemaSupported(s)
```
Comment (parthchandra) [inline spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:1122]: Is this change needed? In general, structs are not supported (yet) and since this method is not operator specific, we probably shouldn't have this here.

File: spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala (hunk 7 above)
Comment (parthchandra) [inline spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:62]: nit: Can we consider renaming this to `allowStruct` to make it explicit that this is only for structs (and not maps and arrays).

File: spark/src/main/scala/org/apache/spark/sql/comet/CometRowToColumnarExec.scala (hunk 1 above)
Comment (parthchandra) [reply spark/src/main/scala/org/apache/spark/sql/comet/CometRowToColumnarExec.scala:75]: > it could be worth thinking about using the Spark reader as a real fallback use case instead of just for testing purposes
It might be worthwhile, though I think we would like to prioritize the complex type support in the reader.

**Review by parthchandra — APPROVED** [review]
lgtm (some minor comments)

File: spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala (hunk 10 above)
Comment (Kimahriman) [reply spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:1122]: See https://github.com/apache/datafusion-comet/pull/731#discussion_r1698972618, it's still needed by shouldApplyRowToColumnar

Comment (parthchandra) [reply spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:1122]: Hmm. I'm concerned that someone might use this method and get unexpected behavior. A comment to explain why this is here would be justified, I think?

Comment (Kimahriman) [reply spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:1122]: Yeah it probably makes sense just to move it into CometRowToColumnarExec

File: spark/src/main/scala/org/apache/spark/sql/comet/CometRowToColumnarExec.scala (hunk 1 above)
Comment (andygrove) [reply spark/src/main/scala/org/apache/spark/sql/comet/CometRowToColumnarExec.scala:75]: We can change the name in a future PR

File: native/spark-expr/src/structs.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Kimahriman) [reply native/spark-expr/src/structs.rs:151]: In this particular case `GetStructField` might not even make sense as a UDF because it's created by the Spark analyzer from an `ExtractValue` expression, which resolves the name to an ordinal in the struct. It'd be odd to use `GetStructField` directly, since you would normally want a nested column by name, and not by index. I think the existing `get_field` UDF in DataFusion covers that use case, but using this PhysicalExpr on the Spark side seems like it makes more sense instead of having to convert the ordinal back into a field name to then use `get_field`

File: spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala (hunk 7 above)
Comment (Kimahriman) [reply spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:62]: It'd be a little odd to have a config for struct type support IMO, since either the code supports it or it doesn't. Maybe you would want to override to disable struct support, but you would still need all the same code to support it as well

File: spark/src/main/scala/org/apache/spark/sql/comet/CometBatchScanExec.scala (hunk 8 above)
Comment (Kimahriman) [reply spark/src/main/scala/org/apache/spark/sql/comet/CometBatchScanExec.scala:159]: I'll try to make a trait to share at least all of the common types supported "everywhere"

File: spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala (hunk 9 above)
Comment (Kimahriman) [reply spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala]: Added

File: spark/src/main/scala/org/apache/spark/sql/comet/CometBatchScanExec.scala (hunk 8 above)
Comment (Kimahriman) [reply spark/src/main/scala/org/apache/spark/sql/comet/CometBatchScanExec.scala:159]: Naming is hard, but something like
```scala
trait DataTypeSupport {
  def isAdditionallySupported(dt: DataType): Boolean = false

  private def isGloballySupported(dt: DataType): Boolean = dt match {
    case BooleanType | ByteType | ShortType | IntegerType | LongType | FloatType | DoubleType |
        BinaryType | StringType | _: DecimalType | DateType | TimestampType =>
      true
    case t: DataType if t.typeName == "timestamp_ntz" => true
    case _ => false
  }

  def isSchemaSupported(struct: StructType): Boolean = {
    struct.fields.map(_.dataType).forall(isTypeSupported)
  }

  def isTypeSupported(dt: DataType): Boolean = {
    if (isGloballySupported(dt) || isAdditionallySupported(dt)) {
      // If complex types are supported, we additionally want to recurse into their children
      dt match {
        case StructType(fields) => fields.map(_.dataType).forall(isTypeSupported)
        case ArrayType(elementType, _) => isTypeSupported(elementType)
        case MapType(keyType, valueType, _) =>
          isTypeSupported(keyType) && isTypeSupported(valueType)
        // Not a complex type
        case _ => true
      }
    } else {
      false
    }
  }
}
```
?

File: spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala (hunk 7 above)
Comment (Kimahriman) [reply spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala]: renamed

File: spark/src/main/scala/org/apache/spark/sql/comet/CometBatchScanExec.scala (hunk 8 above)
Comment (Kimahriman) [reply spark/src/main/scala/org/apache/spark/sql/comet/CometBatchScanExec.scala]: Pushed, let me know how it looks. Not a huge fan of the function names but couldn't come up with anything better. It could maybe be expanded to other operators in the future if they need to opt in or out of certain data types support.

**Comment by andygrove** [comment]
Thanks for addressing the feedback @Kimahriman. Could you fix the merge conflict?

**Comment by Kimahriman** [comment]
> Thanks for addressing the feedback @Kimahriman. Could you fix the merge conflict?

Done

**Review by andygrove — APPROVED** [review]
LGTM. I think this is ready to merge @parthchandra / @kazuyukitanimura ?

**Review by comphead — APPROVED** [review]
lgtm thanks @Kimahriman this PR def brings up the benefit

File: spark/src/main/scala/org/apache/spark/sql/comet/CometRowToColumnarExec.scala (hunk 1 above)
Comment (Kimahriman) [reply spark/src/main/scala/org/apache/spark/sql/comet/CometRowToColumnarExec.scala:78]: I'll write up an issue

**Review by kazuyukitanimura — APPROVED** [review]
Thanks @Kimahriman Would you mind merge with the latest main to resolve the conflict?

**Comment by Kimahriman** [comment]
> Thanks @Kimahriman Would you mind merge with the latest main to resolve the conflict?

Done again!
